app/src/main/java/com/epfl/drawyourpath/community/TournamentModel.kt

```diff
@@ -1,51 +1,249 @@
+        addSource(allTournaments) { tournaments ->
+            value = tournaments.filter { userTournaments.value!!.contains(it.id) && it.startDate <= LocalDateTime.now() && it.endDate >= LocalDateTime.now() }
+        }
+        addSource(userTournaments) { ids ->
+            value = allTournaments.value!!.filter { ids.contains(it.id) && it.startDate <= LocalDateTime.now() && it.endDate >= LocalDateTime.now() }
+        }
+    }
+
+    val startingSoonTournament: LiveData<List<Pair<Tournament, Boolean>>> = MediatorLiveData<List<Pair<Tournament, Boolean>>>().apply {
+        addSource(allTournaments) { tournaments ->
+            value = tournaments.filter { it.startDate > LocalDateTime.now() && it.endDate >= LocalDateTime.now() }.map { Pair(it, userTournaments.value!!.contains(it.id)) }
+        }
+        addSource(userTournaments) { ids ->
+            value = allTournaments.value!!.filter { it.startDate > LocalDateTime.now() && it.endDate >= LocalDateTime.now() }.map { Pair(it, ids.contains(it.id)) }
+        }
+    }
+
+    val discoverTournament: LiveData<List<Tournament>> = MediatorLiveData<List<Tournament>>().apply {
+        addSource(allTournaments) { tournaments ->
+            value = tournaments.filter { !userTournaments.value!!.contains(it.id) && it.startDate <= LocalDateTime.now() && it.endDate >= LocalDateTime.now() }
+        }
+        addSource(userTournaments) { ids ->
+            value = allTournaments.value!!.filter { !ids.contains(it.id) && it.startDate <= LocalDateTime.now() && it.endDate >= LocalDateTime.now() }
+        }
+    }
+
+    private val postOf: MutableLiveData<String?> = MutableLiveData(null)
+
+    private val allPosts: MutableMap<String, LiveData<List<TournamentPost>>> = mutableMapOf()
+
+    val posts: LiveData<List<TournamentPost>> = postOf.switchMap { id ->
+        id?.let { database.getTournamentPosts(it) } ?: MediatorLiveData<List<TournamentPost>>().apply {
+            addSource(userTournaments) { ids ->
+                value = allPosts.filterKeys { ids.contains(it) }.values.mapNotNull { it.value }.flatten()
+            }
+            addSource(allTournamentIds) { tournamentIds ->
+                tournamentIds.forEach { id ->
+                    allPosts.getOrPut(id) {
+                        database.getTournamentPosts(id).also { livedata ->
+                            addSource(livedata) { posts ->
+                                if (userTournaments.value!!.contains(id)) {
+                                    value = value?.filterNot { post -> posts.map { it.postId }.contains(post.postId) }?.plus(posts) ?: posts
+                                }
+                            }
+                        }
+                    }
+                }
+                value = allPosts.filterKeys { userTournaments.value!!.contains(it) }.values.mapNotNull { it.value }.flatten()
+            }
+        }
+    }

```

Comment:
Maybe you can refactor this onto an helper function since it contains a lot of blocks.